File: blackbox/a2a-handoff-quorum-v0.1/evaluator.mjs
import { createHash } from "node:crypto";

export const PROFILE = "meshfleet.a2a.handoff-quorum.v0.1";
const MAX_BYTES = 131072;
const MAX_DEPTH = 64;
const MAX_ACTIONS = 128;
const MAX_VOTERS = 128;
const MAX_WEIGHT = 1000000;
const MAX_TOTAL_WEIGHT = 10000000;

export class ConformanceError extends Error {
  constructor(code, message = code) {
    super(message);
    this.code = code;
  }
}

function fail(code, message) {
  throw new ConformanceError(code, message);
}

function validScalarString(value) {
  if (typeof value !== "string") return false;
  for (let i = 0; i < value.length; i += 1) {
    const unit = value.charCodeAt(i);
    if (unit >= 0xd800 && unit <= 0xdbff) {
      const next = value.charCodeAt(i + 1);
      if (!(next >= 0xdc00 && next <= 0xdfff)) return false;
      i += 1;
    } else if (unit >= 0xdc00 && unit <= 0xdfff) {
      return false;
    }
  }
  return true;
}

function compareCodePoints(a, b) {
  const aa = Array.from(a, (c) => c.codePointAt(0));
  const bb = Array.from(b, (c) => c.codePointAt(0));
  for (let i = 0; i < Math.min(aa.length, bb.length); i += 1) {
    if (aa[i] !== bb[i]) return aa[i] - bb[i];
  }
  return aa.length - bb.length;
}

export function canonical(value) {
  if (value === null) return "null";
  if (value === true) return "true";
  if (value === false) return "false";
  if (typeof value === "number") {
    if (!Number.isSafeInteger(value)) fail("UNSAFE_INTEGER");
    if (Object.is(value, -0)) fail("NON_CANONICAL_INTEGER");
    return String(value);
  }
  if (typeof value === "string") {
    if (!validScalarString(value)) fail("INVALID_UNICODE");
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) return `[${value.map(canonical).join(",")}]`;
  if (typeof value === "object") {
    const keys = Object.keys(value).sort(compareCodePoints);
    return `{${keys.map((key) => `${canonical(key)}:${canonical(value[key])}`).join(",")}}`;
  }
  fail("INVALID_SCENARIO");
}

export function digest(value) {
  return createHash("sha256").update(canonical(value), "utf8").digest("hex");
}

class Parser {
  constructor(text) {
    this.text = text;
    this.i = 0;
  }

  ws() {
    while (this.i < this.text.length && /[\t\n\r ]/.test(this.text[this.i])) this.i += 1;
  }

  value(depth = 0) {
    if (depth > MAX_DEPTH) fail("DEPTH_LIMIT");
    this.ws();
    const c = this.text[this.i];
    if (c === "{") return this.object(depth + 1);
    if (c === "[") return this.array(depth + 1);
    if (c === '"') return this.string();
    if (c === "t" && this.text.slice(this.i, this.i + 4) === "true") {
      this.i += 4;
      return true;
    }
    if (c === "f" && this.text.slice(this.i, this.i + 5) === "false") {
      this.i += 5;
      return false;
    }
    if (c === "n" && this.text.slice(this.i, this.i + 4) === "null") {
      this.i += 4;
      return null;
    }
    if (c === "-" || (c >= "0" && c <= "9")) return this.number();
    fail("MALFORMED_JSON");
  }

  string() {
    const start = this.i;
    this.i += 1;
    let escaped = false;
    while (this.i < this.text.length) {
      const c = this.text[this.i];
      if (!escaped && c === '"') {
        this.i += 1;
        let value;
        try {
          value = JSON.parse(this.text.slice(start, this.i));
        } catch {
          fail("MALFORMED_JSON");
        }
        if (!validScalarString(value)) fail("INVALID_UNICODE");
        return value;
      }
      if (!escaped && c.charCodeAt(0) < 0x20) fail("MALFORMED_JSON");
      if (!escaped && c === "\\") escaped = true;
      else escaped = false;
      this.i += 1;
    }
    fail("MALFORMED_JSON");
  }

  number() {
    const rest = this.text.slice(this.i);
    const match = /^-?(?:0|[1-9]\d*)/.exec(rest);
    if (!match) fail("MALFORMED_JSON");
    const token = match[0];
    this.i += token.length;
    if (token === "-0") fail("NON_CANONICAL_INTEGER");
    const suffix = this.text.slice(this.i);
    if (/^\.\d/.test(suffix) || /^[eE][+-]?\d/.test(suffix)) fail("NON_CANONICAL_INTEGER");
    if (/^[0-9.eE+-]/.test(suffix)) fail("MALFORMED_JSON");
    const value = Number(token);
    if (!Number.isSafeInteger(value)) fail("UNSAFE_INTEGER");
    return value;
  }

  array(depth) {
    this.i += 1;
    const result = [];
    this.ws();
    if (this.text[this.i] === "]") {
      this.i += 1;
      return result;
    }
    while (true) {
      result.push(this.value(depth));
      this.ws();
      if (this.text[this.i] === "]") {
        this.i += 1;
        return result;
      }
      if (this.text[this.i] !== ",") fail("MALFORMED_JSON");
      this.i += 1;
    }
  }

  object(depth) {
    this.i += 1;
    const result = {};
    const seen = new Set();
    this.ws();
    if (this.text[this.i] === "}") {
      this.i += 1;
      return result;
    }
    while (true) {
      this.ws();
      if (this.text[this.i] !== '"') fail("MALFORMED_JSON");
      const key = this.string();
      if (seen.has(key)) fail("DUPLICATE_MEMBER");
      seen.add(key);
      this.ws();
      if (this.text[this.i] !== ":") fail("MALFORMED_JSON");
      this.i += 1;
      result[key] = this.value(depth);
      this.ws();
      if (this.text[this.i] === "}") {
        this.i += 1;
        return result;
      }
      if (this.text[this.i] !== ",") fail("MALFORMED_JSON");
      this.i += 1;
    }
  }
}

export function parseStrictJson(input) {
  let bytes;
  if (typeof input === "string") bytes = Buffer.from(input, "utf8");
  else if (Buffer.isBuffer(input) || input instanceof Uint8Array) bytes = Buffer.from(input);
  else fail("INVALID_SCENARIO");
  if (bytes.length > MAX_BYTES) fail("SIZE_LIMIT");
  let text;
  try {
    text = new TextDecoder("utf-8", { fatal: true }).decode(bytes);
  } catch {
    fail("INVALID_UTF8");
  }
  const parser = new Parser(text);
  const value = parser.value();
  parser.ws();
  if (parser.i !== text.length) fail("MALFORMED_JSON");
  return value;
}

function object(value, code = "INVALID_SCENARIO") {
  if (!value || typeof value !== "object" || Array.isArray(value)) fail(code);
  return value;
}

function exact(value, required, code = "INVALID_SCENARIO") {
  const keys = Object.keys(value);
  for (const key of required) if (!Object.hasOwn(value, key)) fail("MISSING_FIELD", key);
  for (const key of keys) if (!required.includes(key)) fail("UNKNOWN_FIELD", key);
}

function id(value) {
  return typeof value === "string" && value.length > 0 && validScalarString(value);
}

function integer(value) {
  return typeof value === "number" && Number.isSafeInteger(value) && value >= 0;
}

function validateScenario(raw) {
  const scenario = object(raw);
  exact(scenario, ["profile", "case_id", "proposal", "actions"]);
  if (scenario.profile !== PROFILE) fail("PROFILE_REJECT");
  if (!id(scenario.case_id)) fail("INVALID_FIELD");
  const proposal = object(scenario.proposal);
  exact(proposal, [
    "proposal_id", "voters", "required_signoffs", "quorum", "weights", "deadline", "silence_policy",
  ]);
  if (!id(proposal.proposal_id)) fail("INVALID_FIELD");
  if (!Array.isArray(proposal.voters) || proposal.voters.length < 1 || proposal.voters.length > MAX_VOTERS) {
    fail("INVALID_FIELD");
  }
  if (!proposal.voters.every(id)) fail("INVALID_FIELD");
  if (new Set(proposal.voters).size !== proposal.voters.length) fail("DUPLICATE_VOTER");
  if (!Array.isArray(proposal.required_signoffs) || !proposal.required_signoffs.every(id)) {
    fail("INVALID_FIELD");
  }
  if (new Set(proposal.required_signoffs).size !== proposal.required_signoffs.length) fail("INVALID_SIGNOFF");
  if (proposal.required_signoffs.some((v) => !proposal.voters.includes(v))) fail("INVALID_SIGNOFF");
  const weights = object(proposal.weights);
  for (const [voter, weight] of Object.entries(weights)) {
    if (!proposal.voters.includes(voter)) fail("INVALID_WEIGHT");
    if (!Number.isSafeInteger(weight) || weight < 1 || weight > MAX_WEIGHT) fail("INVALID_WEIGHT");
  }
  const weightOf = (voter) => Object.hasOwn(weights, voter) ? weights[voter] : 1;
  const totalWeight = proposal.voters.reduce((sum, voter) => sum + weightOf(voter), 0);
  if (totalWeight > MAX_TOTAL_WEIGHT) fail("INVALID_WEIGHT");
  if (!Number.isSafeInteger(proposal.quorum) || proposal.quorum < 1 || proposal.quorum > totalWeight) {
    fail("INVALID_QUORUM");
  }
  if (proposal.deadline !== null && !integer(proposal.deadline)) fail("INVALID_FIELD");
  if (!["abstain", "approve"].includes(proposal.silence_policy)) fail("INVALID_FIELD");
  if (!Array.isArray(scenario.actions) || scenario.actions.length > MAX_ACTIONS) fail("INVALID_FIELD");
  let previousAt = -1;
  for (const action of scenario.actions) {
    object(action);
    if (action.op === "vote") {
      exact(action, ["op", "at", "receipt_id", "voter_id", "seq", "decision"]);
      if (!integer(action.at) || !integer(action.seq) || !id(action.receipt_id) || !id(action.voter_id)) {
        fail("INVALID_FIELD");
      }
      if (!["approve", "decline"].includes(action.decision)) fail("INVALID_FIELD");
    } else if (action.op === "resolve") {
      exact(action, ["op", "at"]);
      if (!integer(action.at)) fail("INVALID_FIELD");
    } else {
      fail("INVALID_FIELD");
    }
    if (action.at < previousAt) fail("NON_MONOTONIC_TIME");
    previousAt = action.at;
  }
  return scenario;
}

function publicState(state) {
  return {
    proposal: state.proposal,
    status: state.status,
    resolved_at: state.resolved_at,
    votes: state.votes,
    events: state.events,
  };
}

function tally(state, now) {
  const voters = state.proposal.voters;
  const approvals = voters.filter((v) => state.effective.get(v)?.decision === "approve");
  const declines = voters.filter((v) => state.effective.get(v)?.decision === "decline");
  const pending = voters.filter((v) => !state.effective.has(v));
  const weightOf = (v) => Object.hasOwn(state.proposal.weights, v) ? state.proposal.weights[v] : 1;
  const sum = (values) => values.reduce((total, voter) => total + weightOf(voter), 0);
  const approvalWeight = sum(approvals);
  const declineWeight = sum(declines);
  const pendingWeight = sum(pending);
  const signoffsMet = state.proposal.required_signoffs.every(
    (v) => state.effective.get(v)?.decision === "approve",
  );
  const signoffRejected = state.proposal.required_signoffs.some(
    (v) => state.effective.get(v)?.decision === "decline",
  );
  const deadlinePassed = state.proposal.deadline !== null && now >= state.proposal.deadline;
  const effectiveApprovalWeight =
    deadlinePassed && state.proposal.silence_policy === "approve"
      ? approvalWeight + pendingWeight
      : approvalWeight;
  const reachable = approvalWeight + pendingWeight >= state.proposal.quorum && !signoffRejected;
  let status = state.status;
  if (status === "open") {
    if (signoffRejected || !reachable) status = "rejected";
    else if (effectiveApprovalWeight >= state.proposal.quorum && signoffsMet) status = "ratified";
    else if (deadlinePassed) status = "expired";
  }
  return {
    status,
    approvals,
    declines,
    pending,
    required_signoffs: state.proposal.required_signoffs,
    signoffs_met: signoffsMet,
    reachable,
    approval_weight: approvalWeight,
    decline_weight: declineWeight,
    pending_weight: pendingWeight,
    total_weight: approvalWeight + declineWeight + pendingWeight,
  };
}

function event(state, kind, at, detail) {
  state.events.push({ seq: state.events.length, kind, at, ...detail });
}

export function evaluateScenario(raw) {
  const scenario = validateScenario(raw);
  const state = {
    proposal: structuredClone(scenario.proposal),
    status: "open",
    resolved_at: null,
    votes: [],
    events: [],
    effective: new Map(),
    receiptContent: new Map(),
  };
  const commandResults = [];
  for (let index = 0; index < scenario.actions.length; index += 1) {
    const action = scenario.actions[index];
    const pre = digest(publicState(state));
    let accepted = true;
    let outcome = null;
    let error = null;
    if (action.op === "vote") {
      const fingerprint = canonical(action);
      const prior = state.receiptContent.get(action.receipt_id);
      if (prior !== undefined) {
        if (prior === fingerprint) outcome = "idempotent_replay";
        else {
          accepted = false;
          error = "RECEIPT_REPLAY_CONFLICT";
        }
      } else if (state.status !== "open") {
        accepted = false;
        error = "RATIFICATION_TERMINAL";
      } else if (!state.proposal.voters.includes(action.voter_id)) {
        accepted = false;
        error = "UNKNOWN_VOTER";
      } else {
        const current = state.effective.get(action.voter_id);
        const expectedSeq = current ? current.seq + 1 : 0;
        if (action.seq !== expectedSeq) {
          accepted = false;
          error = "VOTE_SEQUENCE";
        } else if (current?.decision === action.decision) {
          accepted = false;
          error = "VOTE_NO_CHANGE";
        } else {
          const vote = {
            receipt_id: action.receipt_id,
            voter_id: action.voter_id,
            seq: action.seq,
            decision: action.decision,
            at: action.at,
          };
          state.votes.push(vote);
          state.effective.set(action.voter_id, vote);
          state.receiptContent.set(action.receipt_id, fingerprint);
          event(state, "vote_recorded", action.at, {
            receipt_id: action.receipt_id,
            voter_id: action.voter_id,
            vote_seq: action.seq,
            decision: action.decision,
          });
          outcome = "vote_recorded";
        }
      }
    } else {
      if (state.status !== "open") {
        outcome = state.status;
      } else {
        const current = tally(state, action.at);
        if (current.status !== "open") {
          state.status = current.status;
          state.resolved_at = action.at;
          event(state, "ratification_resolved", action.at, { status: current.status });
        }
        outcome = current.status;
      }
    }
    const post = digest(publicState(state));
    commandResults.push({
      index,
      op: action.op,
      accepted,
      outcome,
      error,
      pre_state_sha256: pre,
      post_state_sha256: post,
    });
  }
  const now = scenario.actions.length ? scenario.actions.at(-1).at : 0;
  return {
    profile: PROFILE,
    case_id: scenario.case_id,
    status: state.status,
    resolved_at: state.resolved_at,
    tally: tally(state, now),
    votes: state.votes,
    events: state.events,
    command_results: commandResults,
    state_sha256: digest(publicState(state)),
  };
}

export function evaluateBytes(bytes) {
  return evaluateScenario(parseStrictJson(bytes));
}

export function projection(result) {
  return {
    case_id: result.case_id,
    stored_status: result.status,
    tally_status: result.tally.status,
    approvals: result.tally.approvals,
    declines: result.tally.declines,
    pending: result.tally.pending,
    signoffs_met: result.tally.signoffs_met,
    reachable: result.tally.reachable,
    approval_weight: result.tally.approval_weight,
    decline_weight: result.tally.decline_weight,
    pending_weight: result.tally.pending_weight,
    total_weight: result.tally.total_weight,
    accepted_actions: result.command_results.filter((r) => r.accepted).length,
    rejected_actions: result.command_results.filter((r) => !r.accepted).length,
    error_codes: result.command_results.filter((r) => r.error).map((r) => r.error),
    event_kinds: result.events.map((e) => e.kind),
  };
}

File: blackbox/a2a-handoff-quorum-v0.1/python/evaluator.py
from __future__ import annotations

import hashlib
import json
from typing import Any

PROFILE = "meshfleet.a2a.handoff-quorum.v0.1"
MAX_BYTES = 131072
MAX_DEPTH = 64
MAX_ACTIONS = 128
MAX_VOTERS = 128
MAX_WEIGHT = 1000000
MAX_TOTAL_WEIGHT = 10000000
MAX_SAFE = 9007199254740991


class ConformanceError(Exception):
    def __init__(self, code: str, message: str | None = None):
        super().__init__(message or code)
        self.code = code


def fail(code: str, message: str | None = None) -> None:
    raise ConformanceError(code, message)


def valid_scalar_string(value: Any) -> bool:
    if not isinstance(value, str):
        return False
    return not any(0xD800 <= ord(char) <= 0xDFFF for char in value)


def _validate_value(value: Any, depth: int = 0) -> None:
    if depth > MAX_DEPTH:
        fail("DEPTH_LIMIT")
    if value is None or isinstance(value, bool):
        return
    if isinstance(value, int):
        if abs(value) > MAX_SAFE:
            fail("UNSAFE_INTEGER")
        return
    if isinstance(value, str):
        if not valid_scalar_string(value):
            fail("INVALID_UNICODE")
        return
    if isinstance(value, list):
        for item in value:
            _validate_value(item, depth + 1)
        return
    if isinstance(value, dict):
        for key, item in value.items():
            if not valid_scalar_string(key):
                fail("INVALID_UNICODE")
            _validate_value(item, depth + 1)
        return
    fail("INVALID_SCENARIO")


def canonical(value: Any) -> str:
    _validate_value(value)
    if value is None:
        return "null"
    if value is True:
        return "true"
    if value is False:
        return "false"
    if isinstance(value, int):
        return str(value)
    if isinstance(value, str):
        return json.dumps(value, ensure_ascii=False, separators=(",", ":"))
    if isinstance(value, list):
        return "[" + ",".join(canonical(item) for item in value) + "]"
    keys = sorted(value.keys())
    return "{" + ",".join(canonical(key) + ":" + canonical(value[key]) for key in keys) + "}"


def digest(value: Any) -> str:
    return hashlib.sha256(canonical(value).encode("utf-8")).hexdigest()


def parse_strict_json(raw: bytes | str) -> Any:
    if isinstance(raw, str):
        raw = raw.encode("utf-8")
    if not isinstance(raw, bytes):
        fail("INVALID_SCENARIO")
    if len(raw) > MAX_BYTES:
        fail("SIZE_LIMIT")
    try:
        text = raw.decode("utf-8", errors="strict")
    except UnicodeDecodeError:
        fail("INVALID_UTF8")

    def pairs_hook(pairs: list[tuple[str, Any]]) -> dict[str, Any]:
        result: dict[str, Any] = {}
        for key, value in pairs:
            if key in result:
                fail("DUPLICATE_MEMBER")
            result[key] = value
        return result

    def parse_int(token: str) -> int:
        if token == "-0":
            fail("NON_CANONICAL_INTEGER")
        value = int(token)
        if abs(value) > MAX_SAFE:
            fail("UNSAFE_INTEGER")
        return value

    def parse_float(_: str) -> None:
        fail("NON_CANONICAL_INTEGER")

    def parse_constant(_: str) -> None:
        fail("NON_CANONICAL_INTEGER")

    try:
        value = json.loads(
            text,
            object_pairs_hook=pairs_hook,
            parse_int=parse_int,
            parse_float=parse_float,
            parse_constant=parse_constant,
        )
    except ConformanceError:
        raise
    except (json.JSONDecodeError, RecursionError):
        fail("MALFORMED_JSON")
    _validate_value(value)
    return value


def require_object(value: Any, code: str = "INVALID_SCENARIO") -> dict[str, Any]:
    if not isinstance(value, dict):
        fail(code)
    return value


def exact(value: dict[str, Any], required: list[str]) -> None:
    for key in required:
        if key not in value:
            fail("MISSING_FIELD", key)
    for key in value:
        if key not in required:
            fail("UNKNOWN_FIELD", key)


def identifier(value: Any) -> bool:
    return isinstance(value, str) and len(value) > 0 and valid_scalar_string(value)


def integer(value: Any) -> bool:
    return type(value) is int and 0 <= value <= MAX_SAFE


def validate_scenario(raw: Any) -> dict[str, Any]:
    scenario = require_object(raw)
    exact(scenario, ["profile", "case_id", "proposal", "actions"])
    if scenario["profile"] != PROFILE:
        fail("PROFILE_REJECT")
    if not identifier(scenario["case_id"]):
        fail("INVALID_FIELD")
    proposal = require_object(scenario["proposal"])
    exact(proposal, [
        "proposal_id", "voters", "required_signoffs", "quorum", "weights", "deadline", "silence_policy"
    ])
    if not identifier(proposal["proposal_id"]):
        fail("INVALID_FIELD")
    voters = proposal["voters"]
    if not isinstance(voters, list) or not 1 <= len(voters) <= MAX_VOTERS or not all(identifier(v) for v in voters):
        fail("INVALID_FIELD")
    if len(set(voters)) != len(voters):
        fail("DUPLICATE_VOTER")
    signoffs = proposal["required_signoffs"]
    if not isinstance(signoffs, list) or not all(identifier(v) for v in signoffs):
        fail("INVALID_FIELD")
    if len(set(signoffs)) != len(signoffs) or any(v not in voters for v in signoffs):
        fail("INVALID_SIGNOFF")
    weights = require_object(proposal["weights"])
    for voter, weight in weights.items():
        if voter not in voters or type(weight) is not int or not 1 <= weight <= MAX_WEIGHT:
            fail("INVALID_WEIGHT")
    total_weight = sum(weights.get(voter, 1) for voter in voters)
    if total_weight > MAX_TOTAL_WEIGHT:
        fail("INVALID_WEIGHT")
    quorum = proposal["quorum"]
    if type(quorum) is not int or not 1 <= quorum <= total_weight:
        fail("INVALID_QUORUM")
    if proposal["deadline"] is not None and not integer(proposal["deadline"]):
        fail("INVALID_FIELD")
    if proposal["silence_policy"] not in ("abstain", "approve"):
        fail("INVALID_FIELD")
    actions = scenario["actions"]
    if not isinstance(actions, list) or len(actions) > MAX_ACTIONS:
        fail("INVALID_FIELD")
    previous_at = -1
    for action in actions:
        action = require_object(action)
        if action.get("op") == "vote":
            exact(action, ["op", "at", "receipt_id", "voter_id", "seq", "decision"])
            if (
                not integer(action["at"])
                or not integer(action["seq"])
                or not identifier(action["receipt_id"])
                or not identifier(action["voter_id"])
                or action["decision"] not in ("approve", "decline")
            ):
                fail("INVALID_FIELD")
        elif action.get("op") == "resolve":
            exact(action, ["op", "at"])
            if not integer(action["at"]):
                fail("INVALID_FIELD")
        else:
            fail("INVALID_FIELD")
        if action["at"] < previous_at:
            fail("NON_MONOTONIC_TIME")
        previous_at = action["at"]
    return scenario


def public_state(state: dict[str, Any]) -> dict[str, Any]:
    return {
        "proposal": state["proposal"],
        "status": state["status"],
        "resolved_at": state["resolved_at"],
        "votes": state["votes"],
        "events": state["events"],
    }


def tally(state: dict[str, Any], now: int) -> dict[str, Any]:
    proposal = state["proposal"]
    voters = proposal["voters"]
    effective = state["effective"]
    approvals = [v for v in voters if effective.get(v, {}).get("decision") == "approve"]
    declines = [v for v in voters if effective.get(v, {}).get("decision") == "decline"]
    pending = [v for v in voters if v not in effective]

    def weight(voter: str) -> int:
        return proposal["weights"].get(voter, 1)

    approval_weight = sum(weight(v) for v in approvals)
    decline_weight = sum(weight(v) for v in declines)
    pending_weight = sum(weight(v) for v in pending)
    signoffs_met = all(effective.get(v, {}).get("decision") == "approve" for v in proposal["required_signoffs"])
    signoff_rejected = any(effective.get(v, {}).get("decision") == "decline" for v in proposal["required_signoffs"])
    deadline_passed = proposal["deadline"] is not None and now >= proposal["deadline"]
    effective_approval_weight = approval_weight
    if deadline_passed and proposal["silence_policy"] == "approve":
        effective_approval_weight += pending_weight
    reachable = approval_weight + pending_weight >= proposal["quorum"] and not signoff_rejected
    status = state["status"]
    if status == "open":
        if signoff_rejected or not reachable:
            status = "rejected"
        elif effective_approval_weight >= proposal["quorum"] and signoffs_met:
            status = "ratified"
        elif deadline_passed:
            status = "expired"
    return {
        "status": status,
        "approvals": approvals,
        "declines": declines,
        "pending": pending,
        "required_signoffs": proposal["required_signoffs"],
        "signoffs_met": signoffs_met,
        "reachable": reachable,
        "approval_weight": approval_weight,
        "decline_weight": decline_weight,
        "pending_weight": pending_weight,
        "total_weight": approval_weight + decline_weight + pending_weight,
    }


def append_event(state: dict[str, Any], kind: str, at: int, detail: dict[str, Any]) -> None:
    state["events"].append({"seq": len(state["events"]), "kind": kind, "at": at, **detail})


def evaluate_scenario(raw: Any) -> dict[str, Any]:
    scenario = validate_scenario(raw)
    state: dict[str, Any] = {
        "proposal": json.loads(json.dumps(scenario["proposal"], ensure_ascii=False)),
        "status": "open",
        "resolved_at": None,
        "votes": [],
        "events": [],
        "effective": {},
        "receipt_content": {},
    }
    command_results = []
    for index, action in enumerate(scenario["actions"]):
        pre = digest(public_state(state))
        accepted = True
        outcome = None
        error = None
        if action["op"] == "vote":
            fingerprint = canonical(action)
            prior = state["receipt_content"].get(action["receipt_id"])
            if prior is not None:
                if prior == fingerprint:
                    outcome = "idempotent_replay"
                else:
                    accepted = False
                    error = "RECEIPT_REPLAY_CONFLICT"
            elif state["status"] != "open":
                accepted = False
                error = "RATIFICATION_TERMINAL"
            elif action["voter_id"] not in state["proposal"]["voters"]:
                accepted = False
                error = "UNKNOWN_VOTER"
            else:
                current = state["effective"].get(action["voter_id"])
                expected_seq = current["seq"] + 1 if current else 0
                if action["seq"] != expected_seq:
                    accepted = False
                    error = "VOTE_SEQUENCE"
                elif current and current["decision"] == action["decision"]:
                    accepted = False
                    error = "VOTE_NO_CHANGE"
                else:
                    vote = {
                        "receipt_id": action["receipt_id"],
                        "voter_id": action["voter_id"],
                        "seq": action["seq"],
                        "decision": action["decision"],
                        "at": action["at"],
                    }
                    state["votes"].append(vote)
                    state["effective"][action["voter_id"]] = vote
                    state["receipt_content"][action["receipt_id"]] = fingerprint
                    append_event(state, "vote_recorded", action["at"], {
                        "receipt_id": action["receipt_id"],
                        "voter_id": action["voter_id"],
                        "vote_seq": action["seq"],
                        "decision": action["decision"],
                    })
                    outcome = "vote_recorded"
        else:
            if state["status"] != "open":
                outcome = state["status"]
            else:
                current = tally(state, action["at"])
                if current["status"] != "open":
                    state["status"] = current["status"]
                    state["resolved_at"] = action["at"]
                    append_event(state, "ratification_resolved", action["at"], {"status": current["status"]})
                outcome = current["status"]
        post = digest(public_state(state))
        command_results.append({
            "index": index,
            "op": action["op"],
            "accepted": accepted,
            "outcome": outcome,
            "error": error,
            "pre_state_sha256": pre,
            "post_state_sha256": post,
        })
    now = scenario["actions"][-1]["at"] if scenario["actions"] else 0
    return {
        "profile": PROFILE,
        "case_id": scenario["case_id"],
        "status": state["status"],
        "resolved_at": state["resolved_at"],
        "tally": tally(state, now),
        "votes": state["votes"],
        "events": state["events"],
        "command_results": command_results,
        "state_sha256": digest(public_state(state)),
    }


def evaluate_bytes(raw: bytes) -> dict[str, Any]:
    return evaluate_scenario(parse_strict_json(raw))


def projection(result: dict[str, Any]) -> dict[str, Any]:
    tally_value = result["tally"]
    return {
        "case_id": result["case_id"],
        "stored_status": result["status"],
        "tally_status": tally_value["status"],
        "approvals": tally_value["approvals"],
        "declines": tally_value["declines"],
        "pending": tally_value["pending"],
        "signoffs_met": tally_value["signoffs_met"],
        "reachable": tally_value["reachable"],
        "approval_weight": tally_value["approval_weight"],
        "decline_weight": tally_value["decline_weight"],
        "pending_weight": tally_value["pending_weight"],
        "total_weight": tally_value["total_weight"],
        "accepted_actions": sum(1 for item in result["command_results"] if item["accepted"]),
        "rejected_actions": sum(1 for item in result["command_results"] if not item["accepted"]),
        "error_codes": [item["error"] for item in result["command_results"] if item["error"]],
        "event_kinds": [item["kind"] for item in result["events"]],
    }
